File: src/estimator.cpp
#include "libswervedrive/estimator.h"

#include <utility>

namespace swervedrive
{
/**
 * @brief Construct a new Estimator object
 *
 * @param chassis a \sa{Chassis} object specifying the configuration of the chassis to estimate
 * @param init Starting point [x, y, theta] of the chassis in the world frame. Defaults to zero.
 * @param eta_lambda threshold when we assume the algorithm has converged to the correct
 *     icr based on how much lambda has moved between sucessive iterations
 * @param eta_delta threshold below which a starting point for iterative icr iteration is
 *     selected based on the norm between the measured wheel positions and the wheel positions obtained from that
 * starting point
 * @param min_delta_line_search minimum size of the free parameters delta_m and delta_m
 *     to avoid infinate recursion in the line search for the next iteration of the position of lambda
 * @param max_iter_lambda maximum iterations allowed for the iterative icr estimation
 *     algorithm to converge on one point
 * @param singularity_tolerance how close a point must be to be considered to be
 *     'on a structural singularity'
 */
Estimator::Estimator(const Chassis& chassis, Epsilon init, double eta_lambda, double eta_delta,
                     double min_delta_line_search, double max_iter_lambda, double singularity_tolerance)
  : chassis_(chassis)
  , epsilon_(init)
  , eta_lambda_(eta_lambda)
  , eta_delta_(eta_delta)
  , min_delta_line_search_(min_delta_line_search)
  , max_iter_lambda_(max_iter_lambda)
  , singularity_tolerance_(singularity_tolerance){};

/**
 * @brief
 *
 * @param q
 * @return Lambda
 */
Lambda Estimator::estimate(Eigen::VectorXd q)
{
  using namespace Eigen;
  using namespace std;
  using namespace swervedrive;

  std::vector<Lambda> starting_points = select_starting_points(q);

  optional<Lambda> closest_lambda = {};
  optional<double> closest_dist = {};
  for (auto lambda_start : starting_points) {

    Lambda lambda = lambda_start;
    double total_displacement = chassis_.lambda_joint_dist(q, lambda_start);

    // populate closest_lambda if not already
    if (!closest_lambda) {
      closest_lambda = lambda_start;
      closest_dist = chassis_.lambda_joint_dist(q, lambda_start);
    }

    if (chassis_.lambda_joint_dist(q, lambda) < eta_delta_) {
      return lambda;
    } else {
      for (int i=0; i < max_iter_lambda_; ++i) {
        Derivatives d = compute_derivatives(lambda);
        Deltas deltas = solve(d, q, lambda);
        bool diverging;
        Lambda lambda_t = update_parameters(lambda, deltas, q, diverging);
        optional<int> singularity = chassis_.singularity(lambda_t);

        // TODO use the singularity calc
        if (chassis_.lambda_joint_dist(q, lambda_t) > total_displacement) {
          break;
        } else if ((lambda - lambda_t).norm() < eta_lambda_) {
          return lambda_t;
        }
        lambda = lambda_t;
      }
    }
  }
  // If we exhausted all the starting points and couldn't return a value, return the closest
  // TODO closest lambda
}

/**
 * @brief Compute derivatives of constraining surface at lambda.
 *
 * @param lambda Position of ICR.
 * @return Derivatives (u, v, w), the vectors containing the derivatives of each steering angle
 *     with respect to u, v, and w respectively.
 */
Derivatives Estimator::compute_derivatives(Lambda lambda)
{
  using namespace Eigen;
  using namespace std;
  using namespace swervedrive;
  // Work out the best hemisphere to work in
  RowVector3d dm, dn;
  vector<pair<char, double>> dots;
  dots.push_back(make_pair('u', abs(lambda.dot(Vector3d(1, 0, 0)))));
  dots.push_back(make_pair('v', abs(lambda.dot(Vector3d(0, 1, 0)))));
  dots.push_back(make_pair('w', abs(lambda.dot(Vector3d(0, 0, 1)))));
  sort(dots.begin(), dots.end(),
       [](const pair<char, double> p1, const pair<char, double> p2) { return p1.second > p2.second; });
  char axis = dots[0].first;
  if (axis == 'u')
  {
    dm << -lambda(1) / lambda(0), 1, 0;
    dn << -lambda(2) / lambda(0), 0, 1;
  }
  else if (axis == 'v')
  {
    dm << 1, -lambda(0) / lambda(0), 0;
    dn << 0, -lambda(2) / lambda(0), 1;
  }
  else
  {
    dm << 1, 0, -lambda(0) / lambda(0);
    dn << 0, 1, -lambda(1) / lambda(0);
  }

  // we need to elementwise-multiply through broadcasting delta and omega
  // (which are both 1xn row vectors) by three-row matrices, so create
  Eigen::MatrixXd delta_block(3, chassis_.n_);
  Eigen::MatrixXd omega_block(3, chassis_.n_);

  auto diff = chassis_.a_ - chassis_.l_;
  auto delta = lambda.transpose() * diff;
  delta_block << delta, delta, delta;
  auto omega = lambda.transpose() * chassis_.a_orth_;
  omega_block << omega, omega, omega;


  // 3 x n
  MatrixXd gamma_top = (MatrixXd) omega_block.cwiseProduct(diff) + delta_block.cwiseProduct(chassis_.a_orth_);
  // 1 x n
  RowVectorXd gamma_bottom = (RowVectorXd)
      lambda.transpose() * (delta_block.cwiseProduct(diff) - omega_block.cwiseProduct(chassis_.a_orth_));


  RowVectorXd S_m = (dm * gamma_top);
  RowVectorXd S_n = (dn * gamma_top);

  auto singularity = chassis_.singularity(lambda);
  if (singularity) {
    S_m(*singularity) = 0;
    S_n(*singularity) = 0;

    gamma_bottom(*singularity) = 1;
  }

  S_m = S_m.cwiseQuotient(gamma_bottom);
  S_n = S_n.cwiseQuotient(gamma_bottom);

  Derivatives derivatives;
  if (axis == 'u')
  {
    derivatives.u = {};
    derivatives.v = S_m;
    derivatives.w = S_n;
  }
  else if (axis == 'v')
  {
    derivatives.u = S_m;
    derivatives.v = {};
    derivatives.w = S_n;
  }
  else
  {
    derivatives.u = S_m;
    derivatives.v = S_n;
    derivatives.w = {};
  }
  return derivatives;
}

/**
  @brief
  Find the starting points for the Newton-Raphson algorithm. This
  implementation places them at the intersection of the propulsion axis
  and orders them according to their distance to the input point.
  @param q list of angles beta between representing the steer angle
  (measured relative to the orientation orthogonal to the line to the
  chassis frame origin.)
  @returns List of the top three starting points ordered according to
  their distance to the input length.
 */
std::vector<Lambda> Estimator::select_starting_points(Eigen::VectorXd q)
{
  using namespace Eigen;
  using namespace std;
  using namespace swervedrive;

  auto angle_sum = q + chassis_.alpha_;
  auto d = MatrixXd(chassis_.n_, 3);
  d << cos(angle_sum.array()), sin(angle_sum.array()), VectorXd(chassis_.n_);

  vector<Vector3d> p;
  for (int idx = 0; idx < chassis_.n_; ++idx)
  {
    Vector3d si = chassis_.s_.col(idx);
    Vector3d di = d.row(idx);
    auto p_i = si.cross(di).normalized();
    p.push_back(p_i);
  }

  vector<Lambda> starting_points;
  for (int i = 0; i < chassis_.n_; ++i)
  {
    auto p_i = p[i];
    for (int j = i + 1; j < chassis_.n_; ++j)
    {
      auto p_j = p[j];
      if (p_i.isApprox(p_j) || p_i.isApprox(-p_j))
      {
        // Throw away collinear axes
        continue;
      }
      auto c = p_i.cross(p_j).normalized();
      if (c(2) < 0)
      {
        c = -c;
      }
      starting_points.push_back(c);
    }
  }
  // Sort by distance from current configuration
  std::sort(starting_points.begin(), starting_points.end(), [this, q](const Vector3d lambda1, const Vector3d lambda2) {
    auto q1 = chassis_.betas(lambda1);
    auto q2 = chassis_.betas(lambda2);
    return chassis_.displacement(q, q1).array().abs().sum() < chassis_.displacement(q, q2).array().abs().sum();
  });
  return starting_points;
}

/**
 * @brief
 *
 * @param derivatives
 * @param q
 * @param lambda
 * @return Eigen::Vector3d
 */
Deltas Estimator::solve(Derivatives derivatives, Eigen::VectorXd q, Lambda lambda)
{
  using namespace Eigen;

  auto p_zero = chassis_.betas(lambda);
  auto diff = q - p_zero;
  double a_u, a_c, a_v;
  Vector2d b;
  if (!derivatives.u) {
    a_u = (*derivatives.v).dot(*derivatives.v);
    a_c = (*derivatives.v).dot(*derivatives.w);
    a_v = (*derivatives.w).dot(*derivatives.w);
    b << diff.dot(*derivatives.v), diff.dot(*derivatives.w);
  } else if (!derivatives.v) {
    a_u = (*derivatives.u).dot(*derivatives.u);
    a_c = (*derivatives.u).dot(*derivatives.w);
    a_v = (*derivatives.w).dot(*derivatives.w);
    b << diff.dot(*derivatives.u), diff.dot(*derivatives.w);
  } else {
    a_u = (*derivatives.u).dot(*derivatives.u);
    a_c = (*derivatives.u).dot(*derivatives.v);
    a_v = (*derivatives.v).dot(*derivatives.v);
    b << diff.dot(*derivatives.u), diff.dot(*derivatives.v);
  }
  MatrixXd A(2,2);
  A << a_u, a_c, a_c, a_v;
  Vector2d x = A.colPivHouseholderQr().solve(b);
  Deltas d;
  if (!derivatives.u) {
    d.u = {};
    d.v = x(0);
    d.w = x(1);
  } else if (!derivatives.v) {
    d.u = x(0);
    d.v = {};
    d.w = x(1);
  } else {
    d.u = x(0);
    d.v = x(1);
    d.w = {};
  }
  return d;
}

/**
 * @brief
 *
 * @param lambda
 * @param deltas
 * @param q
 * @param diverged
 * @return Lambda
 */
Lambda Estimator::update_parameters(Lambda lambda, Deltas deltas, Eigen::VectorXd q, bool& diverged)
{
  double m, n, delta_m, delta_n;
  std::function<Lambda(double, double)> lambda_t;
  if (!deltas.u) {
    m = lambda(1);
    n = lambda(2);
    delta_m = *deltas.v;
    delta_n = *deltas.w;
    lambda_t = [](double m, double n) {
      double residual = std::max(1 - m*m - n*n, 0.0);
      return Eigen::Vector3d(std::pow(residual, 0.5), m, n).normalized();
    };
  } else if (!deltas.v) {
    m = lambda(0);
    n = lambda(2);
    delta_m = *deltas.u;
    delta_n = *deltas.w;
    lambda_t = [](double m, double n) {
      double residual = std::max(1 - m*m - n*n, 0.0);
      return Eigen::Vector3d(m, std::pow(residual, 0.5), n).normalized();
    };
  } else {
    m = lambda(0);
    n = lambda(1);
    delta_m = *deltas.u;
    delta_n = *deltas.v;
    lambda_t = [](double m, double n) {
      double residual = std::max(1 - m*m - n*n, 0.0);
      return Eigen::Vector3d(m, n, std::pow(residual, 0.5)).normalized();
    };
  }

  double prev_m = m;
  double prev_n = n;
  double total_dist = chassis_.displacement(q, chassis_.betas(lambda)).norm();
  while (1) {
    double m_i = m + delta_m;
    double n_i = n + delta_n;
    // if adding delta_m and delta_m has produced out of bounds values,
    // recursively multiply to ensure they remain within bounds
    while (double factor = std::hypot(m_i, n_i) > 1) {
      m_i /= factor;
      n_i /= factor;
    }
    if (total_dist < chassis_.displacement(q, chassis_.betas(lambda_t(m_i, n_i))).norm()) {
      // Diverging
      // backtrack by reducing the step size
      delta_m *= 0.5;
      delta_n *= 0.5;

      // set a minimum step size to avoid infinite recursion
      if (std::hypot(delta_m, delta_n) < min_delta_line_search_) {
        // Return the previous estimate
        diverged = true;
        return lambda_t(prev_m, prev_n);
      } else {
        prev_m = m_i;
        prev_n = n_i;
      }
    } else {
      diverged = false;
      return lambda_t(m_i, n_i);
    }

  }
}

}  // namespace swervedrive

File: swervedrive/icr/estimator.py
import numpy as np
import math
from .utils import constrain_angle


class Estimator:

    # constants used in the lmda estimation algo
    eta_lmda: float = 1e-4  # TODO: figure out what values this should be
    eta_delta: float = 1e-2  # TODO: figure out what values this should be
    min_delta_size: float = 1e-2  # TODO: figure out what value this should be
    max_iter = 50  # TODO: figure out what value should be
    tolerance: float = 1e-3

    def __init__(
        self,
        epsilon_init: np.ndarray,
        alpha: np.ndarray,
        l: np.ndarray,
    ):
        """
        Initialize the Estimator object. The order in the following arrays
        must be preserved throughout all arguments passed to this object.
        :param epsilon_init: the starting position estimate for the robot
            position. Form (x, y, theta)^T.
        :param alpha: array containing the angle to each of the modules,
            measured counter clockwise from the x-axis.
        :param l: distance to the axis of rotation of each module from
            the origin of the chassis frame
        """

        assert len(alpha.shape) == 2 and alpha.shape[1] == 1, alpha
        assert len(l.shape) == 2 and l.shape[1] == 1, l

        self.epsilon = epsilon_init

        self.alpha = alpha
        self.l = l
        self.n = len(self.alpha)
        self.epsilon_init = epsilon_init

        self.a = np.concatenate([np.cos(alpha).T, np.sin(alpha).T, [[0.0] * self.n]])
        self.a_orth = np.concatenate([-np.sin(alpha).T, np.cos(alpha).T, [[0.0] * self.n]])
        self.s = np.concatenate(
            [(l * np.cos(alpha)).T, (l * np.sin(alpha)).T, [[1.0] * self.n]]
        )
        self.l_v = np.concatenate([[[0.0] * self.n], [[0.0] * self.n], self.l.T])

    def estimate_lmda(self, q: np.ndarray):
        """
        Find the ICR given the steering angles.
        :param q: list of angles beta between representing the steer angle
            (measured relative to the orientation orthogonal to the line to the
            chassis frame origin.)
        :returns: our estimate of ICR as the array (u, v, w)^T.
        """
        assert len(q.shape) == 2 and q.shape[1] == 1, q
        starting_points = self.select_starting_points(q)
        found = False
        closest_lmda = None
        closest_dist = None
        for lmda_start in starting_points:
            lmda = lmda_start
            if closest_lmda is None:
                closest_lmda = lmda_start
                closest_dist = np.linalg.norm(shortest_distance(q, self.S(lmda_start)))
            if np.linalg.norm(shortest_distance(q, self.S(lmda))) < self.eta_delta:
                found = True
            else:
                last_singularity = None
                for i in range(self.max_iter):
                    (S_u, S_v, S_w) = self.compute_derivatives(lmda)
                    if last_singularity is not None:
                        # if we had a singularity last time, set the derivatives
                        # for the corresponding wheel to 0
                        if S_u is not None:
                            S_u[last_singularity] = 0
                        if S_u is not None:
                            S_v[last_singularity] = 0
                        if S_u is not None:
                            S_w[last_singularity] = 0
                    (delta_u, delta_v, delta_w) = self.solve(S_u, S_v, S_w, q, lmda)
                    lmda_t, worse = self.update_parameters(
                        lmda, delta_u, delta_v, delta_w, q
                    )
                    singularity, singularity_number = self.handle_singularities(lmda_t)
                    S_lmda = self.S(lmda_t)
                    if last_singularity is not None and singularity:
                        # the test point is still on the steering axis, suggesting
                        # it is on a singularity. Set beta_k to the input steering
                        # value
                        S_lmda[last_singularity] = q[last_singularity]
                    last_singularity = singularity_number
                    if np.linalg.norm(shortest_distance(q, S_lmda)) > np.linalg.norm(
                        shortest_distance(q, self.S(lmda_start))
                    ):
                        # appears the algorithm has diverged as we are not
                        # improving
                        found = False
                        break
                    else:
                        found = np.linalg.norm(lmda - lmda_t) < self.eta_lmda
                        distance = np.linalg.norm(shortest_distance(q, S_lmda))
                        if distance < closest_dist:
                            closest_lmda = lmda_t
                            closest_dist = distance
                    lmda = lmda_t
                    if found:
                        break
            if found:
                lmda = lmda.reshape(-1, 1)
                # Always return lamdba with a positive w component
                if lmda[2,0] < 0:
                    lmda = -lmda
                return lmda
        lmda = closest_lmda.reshape(-1, 1)
        # Always return lamdba with a positive w component
        if lmda[2,0] < 0:
            lmda = -lmda
        return lmda

    def select_starting_points(self, q: np.ndarray):
        """
        Find the starting points for the Newton-Raphson algorithm. This
        implementation places them at the intersection of the propulsion axis
        and orders them according to their distance to the input point.
        :param q: list of angles beta between representing the steer angle
            (measured relative to the orientation orthogonal to the line to the
            chassis frame origin.)
        :returns: List of the top three starting points ordered according to
        their distance to the input length.
        """
        starting_points = []
        assert len(q.shape) == 2 and q.shape[1] == 1, q

        def get_p(i):
            s = column(self.s, i).reshape(-1)
            d = np.array(
                [
                    math.cos(q[i, 0] + self.alpha[i,0]),
                    math.sin(q[i, 0] + self.alpha[i,0]),
                    0,
                ]
            )
            p = np.cross(s, d)
            p /= np.linalg.norm(p)
            return p

        for i in range(self.n):
            p_1 = get_p(i)
            for j in range(self.n):
                if not i > j:
                    continue
                p_2 = get_p(j)
                c = np.cross(p_1, p_2).reshape(-1, 1)
                if abs(p_1.dot(p_2) / np.linalg.norm(p_1) * np.linalg.norm(p_2)) > 0.99:
                    # the sine of the dot product is zero i.e. they are co-linear:
                    # Throwout cases where the two wheels being compared are co-linear
                    continue
                c /= np.linalg.norm(c)
                if c[2] < 0:
                    c = -c
                dist = np.linalg.norm(shortest_distance(q, self.S(c)))
                starting_points.append([c, dist])
        starting_points.sort(key=lambda point: point[1])
        sp_arr = [p[0].reshape(3, 1) for p in starting_points]
        return sp_arr

    def compute_derivatives(self, lmda: np.ndarray):
        """
        Compute the derivateves of the constraining surface at the current
        estimate of the point.
        :param lmda: position of the ICR estimate
        :returns: np.ndarray with (S_u, S_v, S_w). S_u, S_v, S_w are the vectors
            containing the derivatives of each steering angle in q with respect
            u, v and w respectively.
            One of them will be None because that axis is parameterised in terms
            of the other two.
        """
        assert len(lmda.shape) == 2 and lmda.shape[1] == 1, lmda

        # Work out the best hemisphere to work in
        u = lmda.T.dot(np.array([[1], [0], [0]]))
        v = lmda.T.dot(np.array([[0], [1], [0]]))
        w = lmda.T.dot(np.array([[0], [0], [1]]))
        dots = {'u': abs(u), 'v': abs(v), 'w': abs(w)}
        axis = max(dots.keys(), key=(lambda k: dots[k]))
        if axis == 'u':
            # Parameterise u
            dm = np.array([[-lmda[1,0] / lmda[0,0], 1, 0]])
            dn = np.array([[-lmda[2,0] / lmda[0,0], 0, 1]])
        elif axis == 'v':
            # Parameterise v
            dm = np.array([[1, -lmda[0,0] / lmda[1,0], 0]])
            dn = np.array([[0, -lmda[2,0] / lmda[1,0], 1]])
        else:
            # Parameterise w
            dm = np.array([[1, 0, -lmda[0,0] / lmda[2,0]]])
            dn = np.array([[0, 1, -lmda[1,0] / lmda[2,0]]])

        lmda_T = lmda.T
        lmda_T_block = np.concatenate([lmda_T]*self.n)
        diff = self.a - self.l_v
        delta = lmda_T.dot(diff)
        omega = lmda_T.dot(self.a_orth)
        gamma_top = omega * diff + delta * self.a_orth
        gamma_bottom = lmda_T.dot(delta * diff - omega * self.a_orth)
        lmda_singular = (self.s / np.linalg.norm(self.s, axis=0)).T
        is_singular = np.logical_or(
            np.all(np.isclose(lmda_T_block, lmda_singular, atol=self.tolerance), axis=1),
            np.all(np.isclose(lmda_T_block, -lmda_singular, atol=self.tolerance), axis=1)
        )
        S_m = dm.dot(gamma_top)
        S_n = dn.dot(gamma_top)
        for i,is_sing in enumerate(is_singular):
            if is_sing:
                gamma_bottom[0, i] = 1
                S_m[0, i] = 0
        S_m = (S_m / gamma_bottom).reshape(-1, 1)
        S_n = (S_n / gamma_bottom).reshape(-1, 1)

        if axis == 'u':
            return None, S_m, S_n
        if axis == 'v':
            return S_m, None, S_n
        if axis == 'w':
            return S_m, S_n, None

    def solve(
        self,
        S_u: np.ndarray,
        S_v: np.ndarray,
        S_w: np.ndarray,
        q: np.ndarray,
        lmda: np.ndarray,
    ):
        """
        Solve the system of linear equations to find the free parameters
        delta_u and delta_v.
        :param S_u: derivative of constraining surface wrt u (vector).
        :param S_v: derivative of constraining surface wrt v (vector).
        :param S_v: derivative of constraining surface wrt w (vector).
        :param q: list of angles beta representing the steer angle
        (measured relative to the orientation orthogonal to the line to the
        chassis frame origin.)
        :param lmda: position of the ICR estimate.
        :returns: the free parameters in the form (delta_u, delta_v, delta_w).
        """
        assert len(q.shape) == 2 and q.shape[1] == 1, q
        if S_u is not None:
            assert len(S_u.shape) == 2 and S_u.shape[1] == 1, S_u
        if S_v is not None:
            assert len(S_v.shape) == 2 and S_v.shape[1] == 1, S_v
        if S_w is not None:
            assert len(S_w.shape) == 2 and S_w.shape[1] == 1, S_w
        p_zero = self.S(lmda)
        diff = q - p_zero
        if S_u is None:
            a_u = S_v.T.dot(S_v)
            a_c = S_v.T.dot(S_w)
            a_v = S_w.T.dot(S_w)
            b = np.concatenate((diff.T.dot(S_v), diff.T.dot(S_w)))
        elif S_v is None:
            a_u = S_u.T.dot(S_u)
            a_c = S_u.T.dot(S_w)
            a_v = S_w.T.dot(S_w)
            b = np.concatenate((diff.T.dot(S_u), diff.T.dot(S_w)))
        else:
            a_u = S_u.T.dot(S_u)
            a_c = S_u.T.dot(S_v)
            a_v = S_v.T.dot(S_v)
            b = np.concatenate((diff.T.dot(S_u), diff.T.dot(S_v)))
        A = np.array([
            [a_u[0,0], a_c[0,0]],
            [a_c[0,0], a_v[0,0]]
        ])
        x = np.linalg.solve(A, b)
        if S_u is None:
            return None, x[0, 0], x[1, 0]
        elif S_v is None:
            return x[0, 0], None, x[1, 0]
        else:
            return x[0, 0], x[1, 0], None

    def update_parameters(
        self,
        lmda: np.ndarray,
        delta_u: float,
        delta_v: float,
        delta_w: float,
        q: np.ndarray,
    ):
        """
        Move our estimate of the ICR based on the free parameters delta_u and
        delta_v. If invalid parameters are produced rescale them so they lie
        within the sphere. If the algorithm has diverged backtrack if possible
        :param lmda: current position of the ICR estimate.
        :param delta_u: free parameter defining how much to move the ICR
            estimate in the direction S_u.
        :param delta_v: free parameter defining how much to move the ICR
            estimate in the direction S_v.
        :param delta_w: free parameter defining how much to move the ICR
            estimate in the direction S_w.
        :param q: list of angles beta representing the steer angle
            (measured relative to the orientation orthogonal to the line to the
            chassis frame origin.)
        :returns: the new ICR estimate, a flag indicating divergence of the
            algorithm for this starting point.
        """
        # Move along the non-parameterised axes. Call these m and n
        assert len(lmda.shape) == 2 and lmda.shape[1] == 1, lmda
        assert lmda.shape == (3, 1), lmda
        if delta_u is None:
            m = lmda[1, 0]
            n = lmda[2, 0]
            delta_m = delta_v
            delta_n = delta_w

            def lmda_t(m, n):
                # We normalise here so that we don't get floating point errors
                residual = max(1 - m**2 - n**2, 0.0)
                unnormed = np.array([[residual ** 0.5], [m], [n]])  # Eq 4
                return unnormed / np.linalg.norm(unnormed)

        elif delta_v is None:
            m = lmda[0, 0]
            n = lmda[2, 0]
            delta_m = delta_u
            delta_n = delta_w

            def lmda_t(m, n):
                # We normalise here so that we don't get floating point errors
                residual = max(1 - m**2 - n**2, 0.0)
                unnormed = np.array([[m], [residual ** 0.5], [n]])  # Eq 4
                return unnormed / np.linalg.norm(unnormed)

        else:
            m = lmda[0, 0]
            n = lmda[1, 0]
            delta_m = delta_u
            delta_n = delta_v

            def lmda_t(m, n):
                # We normalise here so that we don't get floating point errors
                residual = max(1 - m**2 - n**2, 0.0)
                unnormed = np.array([[m], [n], [residual ** 0.5]])  # Eq 4
                return unnormed / np.linalg.norm(unnormed)

        prev_m = m
        prev_n = n
        # while the algorithm produces a worse than or equal to good estimate
        # for q on the surface as lmda from the previous iteration
        total_dist = np.linalg.norm(shortest_distance(q, self.S(lmda)))
        while True:
            m_i = m + delta_m
            n_i = n + delta_n
            # if adding delta_m and delta_m has produced out of bounds values,
            # recursively multiply to ensure they remain within bounds
            while np.linalg.norm([m_i, n_i]) > 1:
                factor = np.linalg.norm([m_i, n_i])
                m_i /= factor
                n_i /= factor
            if total_dist < np.linalg.norm(
                shortest_distance(q, self.S(lmda_t(m_i, n_i)))
            ):
                # Diverging
                # backtrack by reducing the step size
                delta_m *= 0.5
                delta_n *= 0.5

                # set a minimum step size to avoid infinite recursion
                if np.linalg.norm([delta_m, delta_n]) < self.min_delta_size:
                    # Return the previous estimate
                    return lmda_t(prev_m, prev_n), True
                else:
                    prev_m = m_i
                    prev_n = n_i
            else:
                return lmda_t(m_i, n_i), False

    def handle_singularities(self, lmda: np.ndarray):
        """
        Handle the structural singularities that may have been produced when
        the parameters were updated (when the ICR lies on a steering axis).
        :param lmda: the ICR estimate after the parameters were updated.
        :returns: if the ICR is on a structural singularity, and the wheel
            number which the singularity is on if there is one
        """
        assert len(lmda.shape) == 2 and lmda.shape[1] == 1, lmda
        wheel_number = None
        for i in range(self.n):
            # equations 16 and 17 in the paper
            s = column(self.s, i)
            if np.allclose(lmda, s / np.linalg.norm(s)):
                wheel_number = i
                break
        return wheel_number is not None, wheel_number

    def S(self, lmda: np.ndarray):
        """
        Compute the point in the joint space (space of all beta steering angle
        values) associated with a particular ICR.
        :param lmda: the ICR to compute the point for.
        :returns: row vector expressing the point.
        """
        assert len(lmda.shape) == 2 and lmda.shape[1] == 1, lmda

        lmda_T = lmda.T
        S = np.arctan2(lmda_T.dot(self.a_orth), lmda_T.dot(self.a - self.l_v))
        S[np.isnan(S)] = math.pi / 2
        return S.reshape(-1, 1)


def shortest_distance(
    q_e: np.ndarray, q_d: np.ndarray, beta_bounds: np.ndarray = None
):
    """
    Determine if the rotation each wheel needs to make to get to the target.
    Respect the joint limits, but allow movement to a position pi from the target position
    if joint limits allow it.
    :param q_e: an array representing all of the current beta angles
    :parem q_d: an array of all the desired beta angles
    :returns: an array of the same length as the input arrays with each component
        as the correct distance of q_e to q_d.
    """
    assert len(q_e.shape) == 2 and q_e.shape[1] == 1, q_e
    assert len(q_d.shape) == 2 and q_d.shape[1] == 1, q_d

    # TODO: make this work if beta bounds is not None
    diff_opp_diff = np.concatenate([constrain_angle(q_d - q_e), constrain_angle(q_d + math.pi - q_e)], axis=1)
    choose = np.argmin(np.absolute(diff_opp_diff), axis=1).reshape(-1, 1)
    return np.take_along_axis(diff_opp_diff, choose, axis=1)


def column(mat, row_i):
    """
    Grab a column from a vector as a numpy column vector.
    :param row_i: row index
    :returns: the column vector (shape (n, 1))
    """
    return mat[:, row_i : row_i + 1]
